How many late days are you using for this assignment? 

In [43]:
#I will be using 0 late days for this assignment
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [44]:
# Load the given datasets
numerical = pd.read_csv("data/chronic_kidney_disease_numerical.csv")
categorical = pd.read_csv("data/chronic_kidney_disease_categorical.csv")
# Print the data
print(numerical)
categorical

     unique_id   age     bp    bgr     bu   sc    sod  pot  hemo   pcv  \
0       203694  38.0   80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0   
1       938027  43.0   60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0   
2       421471  37.0   60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0   
3       764115  70.0   90.0  144.0  125.0  4.0  136.0  4.6  12.0  37.0   
4       240975  47.0   80.0   95.0   35.0  0.9  140.0  4.1   NaN   NaN   
..         ...   ...    ...    ...    ...  ...    ...  ...   ...   ...   
415     522912  56.0   80.0  165.0   55.0  1.8    NaN  NaN  13.5  40.0   
416     475200  64.0  100.0  163.0   54.0  7.2  140.0  4.6   7.9  26.0   
417     168242  48.0   60.0  112.0   44.0  1.2  142.0  4.9  14.5  44.0   
418     474407  21.0   90.0  107.0   40.0  1.7  125.0  3.5   8.3  23.0   
419     926033  50.0   70.0  230.0   50.0  2.2    NaN  NaN  12.0  41.0   

        wbcc  rbcc  
0     7300.0   6.4  
1     7200.0   5.5  
2     5500.0   5.7  
3     8200.0   4.5  
4     

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
425,769512,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
426,872034,2.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,yes,poor,no,no,ckd
427,643007,NaN,NaN,NaN,normal,notpresent,notpresent,yes,yes,yes,poor,no,no,ckd
428,840392,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,no,no,poor,no,no,ckd


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The data in this dataset is that of various patients in a hospital. The features include biomarkers, nutrient counts, gender, age, and data about diseases or conditions the patients may have. The labels in this data are ckd, and notckd, which are used to indicate whether the patients have chronic kidney disease. The features in the given data sets are both categorical and numerical. Categorical data columns include  rbc - red blood cells, htn - hypertension, and appet - appetite. Numerical data columns include bp - blood pressure, sod - sodium, and age.

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [45]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
num_ids = numerical["unique_id"]
num_duplicates = num_ids.duplicated().sum()
print("Number of duplicates in numerical dataset: ")
print(num_duplicates)
# Drop these duplicate rows
numerical = numerical.drop_duplicates(subset="unique_id", keep='first', inplace=False)

# Repeat the same for categorical dataset. Print the duplicate rows and drop them
cat_ids = categorical["unique_id"]
cat_duplicates = cat_ids.duplicated().sum()
print("Number of duplicates in categorical dataset: ")
print(cat_duplicates)

categorical = categorical.drop_duplicates(subset="unique_id", keep='first', inplace=False)




Number of duplicates in numerical dataset: 
30
Number of duplicates in categorical dataset: 
30


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: The rows that are missing from one of the data sets will have NaN in columns for which they do not have data for.

In [46]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
data = pd.merge(numerical, categorical, how = 'outer')
#Print the combined dataset
data

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,103570,24.0,80.0,125.0,NaN,NaN,136.0,3.5,15.4,43.0,5600.0,4.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,104675,65.0,80.0,215.0,133.0,2.5,NaN,NaN,13.2,41.0,NaN,NaN,2.0,1.0,normal,normal,present,notpresent,no,yes,no,good,no,no,ckd
3,105923,90.0,90.0,139.0,89.0,3.0,140.0,4.1,12.0,37.0,7900.0,3.9,1.0,0.0,NaN,normal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
4,107297,60.0,70.0,NaN,NaN,NaN,NaN,NaN,16.4,43.0,10800.0,5.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
395,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,11500.0,3.3,4.0,3.0,normal,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd
396,996076,57.0,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,10400.0,6.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
397,996412,70.0,60.0,120.0,40.0,0.5,140.0,4.6,16.0,43.0,4500.0,4.9,NaN,NaN,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
398,997629,35.0,NaN,99.0,30.0,0.5,135.0,4.9,15.4,48.0,5000.0,5.2,0.0,0.0,normal,normal,NaN,NaN,no,no,no,good,no,no,notckd


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [47]:
# Calculate the percentage of rows that contain atleast one missing value
percentage = 100 * (data.isna().any(axis=1).sum() / len(data))
# Print %
print("Percentage of rows with at least one missing value")
print(percentage)
# Drop these rows from the dataset
data = pd.DataFrame.dropna(data).reset_index(drop=True)
# Print the Dataset
data

Percentage of rows with at least one missing value
61.75000000000001


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,109053,64.0,60.0,106.0,27.0,0.7,150.0,3.3,14.4,42.0,8100.0,4.7,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
2,114717,41.0,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,9100.0,5.2,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
3,118191,42.0,70.0,93.0,32.0,0.9,143.0,4.7,16.6,43.0,7100.0,5.3,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
4,118596,74.0,60.0,88.0,50.0,0.6,147.0,3.7,17.2,53.0,6000.0,4.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,980291,62.0,70.0,122.0,42.0,1.7,136.0,4.7,12.6,39.0,7900.0,3.9,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,yes,no,good,no,no,ckd
149,991031,30.0,80.0,89.0,42.0,0.5,139.0,5.0,16.7,52.0,10200.0,5.0,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,994377,71.0,70.0,81.0,18.0,0.8,145.0,5.0,14.7,44.0,9800.0,6.0,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,995177,69.0,70.0,214.0,96.0,6.3,120.0,3.9,9.4,28.0,11500.0,3.3,4.0,3.0,normal,abnormal,present,present,yes,yes,yes,good,yes,yes,ckd


## Part 3.4: Sort the dataset according to the Labels

In [48]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
data = data.sort_values(by="Target").reset_index(drop=True)
# Print the dataset
data

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,102677,59.0,70.0,76.0,186.0,15.0,135.0,7.6,7.1,22.0,3800.0,2.1,3.0,0.0,normal,abnormal,notpresent,notpresent,yes,no,no,poor,yes,yes,ckd
1,337834,48.0,70.0,117.0,56.0,3.8,111.0,2.5,11.2,32.0,6700.0,3.9,4.0,0.0,normal,abnormal,present,notpresent,yes,no,no,poor,yes,yes,ckd
2,343710,61.0,80.0,173.0,148.0,3.9,135.0,5.2,7.7,24.0,9200.0,3.2,2.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,poor,yes,yes,ckd
3,349892,59.0,70.0,424.0,55.0,1.7,138.0,4.5,12.6,37.0,10200.0,4.1,1.0,3.0,abnormal,abnormal,notpresent,notpresent,yes,yes,yes,good,no,no,ckd
4,353025,40.0,70.0,253.0,150.0,11.9,132.0,5.6,10.9,31.0,8800.0,3.4,3.0,4.0,normal,normal,notpresent,notpresent,yes,yes,no,poor,yes,no,ckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20.0,70.0,123.0,44.0,1.0,135.0,3.8,14.6,44.0,5500.0,4.8,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
149,203466,22.0,60.0,97.0,18.0,1.2,138.0,4.3,13.5,42.0,7900.0,6.4,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
150,482105,44.0,60.0,95.0,46.0,0.5,138.0,4.2,15.0,50.0,7700.0,6.3,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
151,228849,58.0,80.0,100.0,50.0,1.2,140.0,3.5,14.0,50.0,6700.0,6.5,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [49]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']

data = pd.get_dummies(data, columns = categorical_columns)
data = data.astype(int)

# Print the dataset
data

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,102677,59,70,76,186,15,135,7,7,22,3800,2,3,0,0,1,1,0,1,0,1,0,0,1,1,0,1,0,0,1,0,1,0,1,1,0
1,337834,48,70,117,56,3,111,2,11,32,6700,3,4,0,0,1,1,0,0,1,1,0,0,1,1,0,1,0,0,1,0,1,0,1,1,0
2,343710,61,80,173,148,3,135,5,7,24,9200,3,2,0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
3,349892,59,70,424,55,1,138,4,12,37,10200,4,1,3,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
4,353025,40,70,253,150,11,132,5,10,31,8800,3,3,4,0,1,0,1,1,0,1,0,0,1,0,1,1,0,0,1,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
148,486397,20,70,123,44,1,135,3,14,44,5500,4,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
149,203466,22,60,97,18,1,138,4,13,42,7900,6,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
150,482105,44,60,95,46,0,138,4,15,50,7700,6,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
151,228849,58,80,100,50,1,140,3,14,50,6700,6,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

Answer: Using a single column for the binary variable will work equivalently, as a 0 would indicate a no and a 1 would indicate a yes. However, for categorical variables with more than two values this wouldn't be sufficient, since just a 1 or a 0 can't indicate whether one, two, or all condiments are in a sandwich. Thus we would have to separate the categories out as above, so that each condiment has it's own column. From there we could use a 0 or 1 to indicate whether or not the condiment is in the sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [50]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
lower_bound_cutoff = data[numerical_columns].mean() - 3*data[numerical_columns].std()
upper_bound_cutoff = data[numerical_columns].mean() + 3*data[numerical_columns].std()
filter = (data[numerical_columns] < upper_bound_cutoff) & (data[numerical_columns] > lower_bound_cutoff)
print(filter)
filtered_data = data[filter.all(axis=1)].reset_index(drop=True)
# Print the dataset
filtered_data

      age    bp    bgr    bu     sc    sod   pot  hemo   pcv  wbcc  rbcc
0    True  True   True  True  False   True  True  True  True  True  True
1    True  True   True  True   True  False  True  True  True  True  True
2    True  True   True  True   True   True  True  True  True  True  True
3    True  True  False  True   True   True  True  True  True  True  True
4    True  True   True  True  False   True  True  True  True  True  True
..    ...   ...    ...   ...    ...    ...   ...   ...   ...   ...   ...
148  True  True   True  True   True   True  True  True  True  True  True
149  True  True   True  True   True   True  True  True  True  True  True
150  True  True   True  True   True   True  True  True  True  True  True
151  True  True   True  True   True   True  True  True  True  True  True
152  True  True   True  True   True   True  True  True  True  True  True

[153 rows x 11 columns]


,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,343710,61,80,173,148,3,135,5,7,24,9200,3,2,0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
1,397388,63,100,280,35,3,143,3,13,40,9800,4,2,2,0,1,0,1,1,0,0,1,0,1,1,0,0,1,1,0,1,0,1,0,1,0
2,438182,73,80,253,142,4,138,5,10,33,7200,4,2,0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
3,474407,21,90,107,40,1,125,3,8,23,12400,3,4,0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1,0
4,475200,64,100,163,54,7,140,4,7,26,7500,3,4,2,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
129,486397,20,70,123,44,1,135,3,14,44,5500,4,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
130,203466,22,60,97,18,1,138,4,13,42,7900,6,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
131,482105,44,60,95,46,0,138,4,15,50,7700,6,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
132,228849,58,80,100,50,1,140,3,14,50,6700,6,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [51]:
# Normalize the all Numerical Attributes in the dataset.
filtered_data[numerical_columns] = ((filtered_data[numerical_columns] - filtered_data[numerical_columns].mean()) / (filtered_data[numerical_columns].std()))
# Print the dataset
filtered_data

,unique_id,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,343710,0.796206,0.723662,1.137875,3.951101,1.324211,-0.934242,1.472513,-2.795758,-2.712933,0.506560,-1.790290,2,0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
1,397388,0.921652,2.765149,3.506466,-0.236745,1.324211,0.456741,-1.090042,-0.360450,-0.528732,0.770681,-0.679647,2,2,0,1,0,1,1,0,0,1,0,1,1,0,0,1,1,0,1,0,1,0,1,0
2,438182,1.548881,0.723662,2.908784,3.728738,1.974189,-0.412624,1.472513,-1.578104,-1.484320,-0.373843,-0.679647,2,0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
3,474407,-1.712709,1.744405,-0.323125,-0.051442,0.024253,-2.672971,-1.090042,-2.389874,-2.849446,1.915205,-1.790290,4,0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1,0
4,475200,0.984375,2.765149,0.916512,0.467406,3.924126,-0.064878,0.191235,-2.795758,-2.439908,-0.241782,-1.790290,4,2,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
129,486397,-1.775432,-0.297082,0.031057,0.096800,0.024253,-0.934242,-1.090042,0.045435,0.017319,-1.122185,-0.679647,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
130,203466,-1.649986,-1.317826,-0.544488,-0.866775,0.024253,-0.412624,0.191235,-0.360450,-0.255706,-0.065702,1.541638,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
131,482105,-0.270083,-1.317826,-0.588761,0.170921,-0.625726,-0.412624,0.191235,0.451320,0.836394,-0.153742,1.541638,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
132,228849,0.608037,0.723662,-0.478079,0.319164,0.024253,-0.064878,-1.090042,0.045435,0.836394,-0.593943,1.541638,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: The only column in this data set which isn't appropriate for modeling and predictions is the unique_id. The unique_id is just a randomly assigned number for each patient that serves as a unique identifier. It has no relation to the patient in any way. 

In [52]:
#Remove that column
final_data = filtered_data.drop('unique_id', axis=1)
# Print the dataset
final_data

,age,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc,al,su,rbc_abnormal,rbc_normal,pc_abnormal,pc_normal,pcc_notpresent,pcc_present,ba_notpresent,ba_present,htn_no,htn_yes,dm_no,dm_yes,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.796206,0.723662,1.137875,3.951101,1.324211,-0.934242,1.472513,-2.795758,-2.712933,0.506560,-1.790290,2,0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
1,0.921652,2.765149,3.506466,-0.236745,1.324211,0.456741,-1.090042,-0.360450,-0.528732,0.770681,-0.679647,2,2,0,1,0,1,1,0,0,1,0,1,1,0,0,1,1,0,1,0,1,0,1,0
2,1.548881,0.723662,2.908784,3.728738,1.974189,-0.412624,1.472513,-1.578104,-1.484320,-0.373843,-0.679647,2,0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
3,-1.712709,1.744405,-0.323125,-0.051442,0.024253,-2.672971,-1.090042,-2.389874,-2.849446,1.915205,-1.790290,4,0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1,0
4,0.984375,2.765149,0.916512,0.467406,3.924126,-0.064878,0.191235,-2.795758,-2.439908,-0.241782,-1.790290,4,2,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
129,-1.775432,-0.297082,0.031057,0.096800,0.024253,-0.934242,-1.090042,0.045435,0.017319,-1.122185,-0.679647,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
130,-1.649986,-1.317826,-0.544488,-0.866775,0.024253,-0.412624,0.191235,-0.360450,-0.255706,-0.065702,1.541638,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
131,-0.270083,-1.317826,-0.588761,0.170921,-0.625726,-0.412624,0.191235,0.451320,0.836394,-0.153742,1.541638,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
132,0.608037,0.723662,-0.478079,0.319164,0.024253,-0.064878,-1.090042,0.045435,0.836394,-0.593943,1.541638,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [53]:
# Export the dataframe to a new csv file
final_data.to_csv('adjusted_data.csv')
# Export the dataframe to a new json file
final_data.to_json('adjusted_data.json')
pd.set_option('display.max_columns', None)
print(final_data.head(15))

         age        bp       bgr        bu        sc       sod       pot  \
0   0.796206  0.723662  1.137875  3.951101  1.324211 -0.934242  1.472513   
1   0.921652  2.765149  3.506466 -0.236745  1.324211  0.456741 -1.090042   
2   1.548881  0.723662  2.908784  3.728738  1.974189 -0.412624  1.472513   
3  -1.712709  1.744405 -0.323125 -0.051442  0.024253 -2.672971 -1.090042   
4   0.984375  2.765149  0.916512  0.467406  3.924126 -0.064878  0.191235   
5   0.733483 -1.317826  3.683557 -0.199685  0.024253 -1.803607 -1.090042   
6   0.482592  1.744405  0.163875  2.431617  3.274147 -1.629734  0.191235   
7   1.297989 -0.297082  2.045466  2.023951  3.274147 -3.542336 -1.090042   
8  -0.144637 -1.317826  0.916512  1.875708  1.324211  0.108995  0.191235   
9   0.858929 -0.297082  0.008921  0.022679  0.024253 -0.760370  0.191235   
10  0.419869  0.723662  2.045466  1.171557  1.324211 -0.586497  0.191235   
11  0.984375 -1.317826  2.598875  0.615648  1.974189 -0.586497  1.472513   
12  1.423435

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

### Paste the markdown table here

|    |      age |        bp |        bgr |        bu |        sc |        sod |        pot |       hemo |       pcv |      wbcc |      rbcc | al | su | rbc_abnormal | rbc_normal | pc_abnormal | pc_normal | pcc_notpresent | pcc_present | ba_notpresent | ba_present | htn_no | htn_yes | dm_no | dm_yes | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|----|----------|-----------|------------|-----------|-----------|------------|------------|------------|-----------|------------|------------|----|----|---------------|-------------|-------------|-----------|----------------|-------------|---------------|------------|--------|---------|-------|--------|--------|---------|-------------|-------------|-------|--------|--------|---------|-------------|-----------------|
| 0  | 0.796206 | 0.723662  | 1.137875   | 3.951101  | 1.324211  | -0.934242  | 1.472513   | -2.795758  | -2.712933 | 0.506560   | -1.790290  | 2  | 0  | 1             | 0           | 1           | 0         | 1              | 0           | 1             | 0          | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 1           | 0     | 1      | 0      | 1       | 1           | 0               |
| 1  | 0.921652 | 2.765149  | 3.506466   | -0.236745 | 1.324211  | 0.456741   | -1.090042  | -0.360450  | -0.528732 | 0.770681   | -0.679647  | 2  | 2  | 0             | 1           | 0           | 1         | 1              | 0           | 0             | 1          | 0      | 1       | 1     | 0      | 0      | 1       | 1           | 0           | 1     | 0      | 1      | 0       | 1           | 0               |
| 2  | 1.548881 | 0.723662  | 2.908784   | 3.728738  | 1.974189  | -0.412624  | 1.472513   | -1.578104  | -1.484320 | -0.373843  | -0.679647  | 2  | 0  | 1             | 0           | 1           | 0         | 1              | 0           | 1             | 0          | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 1           | 0     | 1      | 0      | 1       | 1           | 0               |
| 3  | -1.712709| 1.744405  | -0.323125  | -0.051442 | 0.024253  | -2.672971  | -1.090042  | -2.389874  | -2.849446 | 1.915205   | -1.790290  | 4  | 0  | 0             | 1           | 1           | 0         | 0              | 1           | 0             | 1          | 1      | 0       | 1     | 0      | 1      | 0       | 1           | 0           | 0     | 1      | 0      | 1       | 1           | 0               |
| 4  | 0.984375 | 2.765149  | 0.916512   | 0.467406  | 3.924126  | -0.064878  | 0.191235   | -2.795758  | -2.439908 | -0.241782  | -1.790290  | 4  | 2  | 1             | 0           | 1           | 0         | 1              | 0           | 0             | 1          | 0      | 1       | 0     | 1      | 1      | 0       | 1           | 0           | 1     | 0      | 1      | 0       | 1           | 0               |
| 5  | 0.733483 | -1.317826 | 3.683557   | -0.199685 | 0.024253  | -1.803607  | -1.090042  | -2.795758  | -2.576421 | 3.147769   | -1.790290  | 3  | 1  | 0             | 1           | 1           | 0         | 0              | 1           | 0             | 1          | 0      | 1       | 1     | 0      | 1      | 0       | 0           | 1           | 0     | 1      | 0      | 1       | 1           | 0               |
| 6  | 0.482592 | 1.744405  | 0.163875   | 2.431617  | 3.274147  | -1.629734  | 0.191235   | -1.983989  | -2.030370 | -0.726004  | -1.790290  | 2  | 0  | 1             | 0           | 1           | 0         | 1              | 0           | 1             | 0          | 0      | 1       | 1     | 0      | 1      | 0       | 1           | 0           | 0     | 1      | 0      | 1       | 1           | 0               |
| 7  | 1.297989 | -0.297082 | 2.045466   | 2.023951  | 3.274147  | -3.542336  | -1.090042  | -1.983989  | -2.166883 | 1.519023   | -1.790290  | 4  | 3  | 0             | 1           | 1           | 0         | 0              | 1           | 0             | 1          | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 1           | 0     | 1      | 0      | 1       | 1           | 0               |
| 8  | -0.144637| -1.317826 | 0.916512   | 1.875708  | 1.324211  | 0.108995   | 0.191235   | -1.983989  | -2.166883 | 2.883648   | -1.790290  | 1  | 0  | 0             | 1           | 0           | 1         | 1              | 0           | 1             | 0          | 0      | 1       | 0     | 1      | 1      | 0       | 1           | 0           | 1     | 0      | 1      | 0       | 1           | 0               |
| 9  | 0.858929 | -0.297082 | 0.008921   | 0.022679  | 0.024253  | -0.760370  | 0.191235   | -0.766335  | -0.665244 | -0.065702  | -1.790290  | 3  | 0  | 0             | 1           | 1           | 0         | 1              | 0           | 1             | 0          | 0      | 1       | 0     | 1      | 1      | 0       | 1           | 0           | 1     | 0      | 1      | 0       | 1           | 0               |
| 10 | 0.419869 | 0.723662  | 2.045466   | 1.171557  | 1.324211  | -0.586497  | 0.191235   | -1.578104  | -1.347807 | -0.285802  | -1.790290  | 3  | 1  | 0             | 1           | 1           | 0         | 0              | 1           | 0             | 1          | 0      | 1       | 0     | 1      | 1      | 0       | 0           | 1           | 0     | 1      | 0      | 1       | 1           | 0               |
| 11 | 0.984375 | -1.317826 | 2.598875   | 0.615648  | 1.974189  | -0.586497  | 1.472513   | -1.983989  | -2.030370 | -0.241782  | -1.790290  | 4  | 1  | 1             | 0           | 1           | 0         | 1              | 0           | 0             | 1          | 0      | 1       | 0     | 1      | 1      | 0       | 1           | 0           | 1     | 0      | 1      | 0       | 1           | 0               |
| 12 | 1.423435 | -1.317826 | -0.079625  | 3.098708  | 2.624168  | -0.760370  | 0.191235   | -1.172219  | -1.211295 | 3.147769   | -0.679647  | 4  | 0  | 0             | 1           | 0           | 1         | 1              | 0           | 1             | 0          | 0      | 1       | 0     | 1      | 1      | 0       | 0           | 1           | 0     | 1      | 0      | 1       | 1           | 0               |
| 13 | 1.423435 | -0.297082 | 2.156148   | 1.505102  | 1.324211  | -1.281988  | 0.191235   | -1.578104  | -1.484320 | -1.078165  | -1.790290  | 3  | 0  | 0             | 1           | 1           | 0         | 0              | 1           | 0             | 1          | 0      | 1       | 0     | 1      | 0      | 1       | 0           | 1           | 0     | 1      | 0      | 1       | 1           | 0               |
| 14 | 1.423435 | -1.317826 | 4.015603   | -0.236745 | 0.024253  | -1.090042  | -1.090042  | -1.578104  | -1.211295 | 1.254903   | -0.679647  | 1  | 0  | 1             | 0           | 0           | 1         | 1              | 0           | 1             | 0          | 1      | 0       | 0     | 1      | 1      | 0       | 0           | 1           | 1     | 0      | 1      | 0       | 1           | 0               |



** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

There are a few ways. First is to break it down into smaller chunks. This will decrease the likelihood that the LLM will hallucinate during the process. Another way is to ask the model to double check it's work, which may lead to it correcting any hallucinations.

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [54]:
# Code to rename all the columns in the dataset
final_data.rename(columns = {
    'age': 'Age',
    'bp': 'Blood Pressure',
    'sg': 'Specific Gravity',
    'al': 'Albumin',
    'su': 'Sugar',
    'rbc_abnormal': 'Red Blood Cells (Abnormal)',
    'rbc_normal': 'Red Blood Cells (Normal)',
    'pc_abnormal': 'Pus Cell (Abnormal)',
    'pc_normal': 'Pus Cell (Normal)',
    'pcc_notpresent': 'Pus Cell Clumps (Not Present)',
    'pcc_present': 'Pus Cell Clumps (Present)',
    'ba_notpresent': 'Bacteria (Not Present)',
    'ba_present': 'Bacteria (Present)',
    'bgr': 'Blood Glucose Random',
    'bu': 'Blood Urea',
    'sc': 'Serum Creatinine',
    'sod': 'Sodium',
    'pot': 'Potassium',
    'hemo': 'Hemoglobin',
    'pcv': 'Packed Cell Volume',
    'wbcc': 'White Blood Cell Count',
    'rbcc': 'Red Blood Cell Count',
    'htn_no': 'Hypertension (No)',
    'htn_yes': 'Hypertension (Yes)',
    'dm_no': 'Diabetes Mellitus (No)',
    'dm_yes': 'Diabetes Mellitus (Yes)',
    'cad_no': 'Coronary Artery Disease (No)',
    'cad_yes': 'Coronary Artery Disease (Yes)',
    'appet_good': 'Appetite (Good)',
    'appet_poor': 'Appetite (Poor)',
    'pe_no': 'Pedal Edema (No)',
    'pe_yes': 'Pedal Edema (Yes)',
    'ane_no': 'Anemia (No)',
    'ane_yes': 'Anemia (Yes)',
    'Target_ckd': 'Target',
    'Target_notckd': 'Target (Not CKD)',
}, inplace = True)


final_data

,Age,Blood Pressure,Blood Glucose Random,Blood Urea,Serum Creatinine,Sodium,Potassium,Hemoglobin,Packed Cell Volume,White Blood Cell Count,Red Blood Cell Count,Albumin,Sugar,Red Blood Cells (Abnormal),Red Blood Cells (Normal),Pus Cell (Abnormal),Pus Cell (Normal),Pus Cell Clumps (Not Present),Pus Cell Clumps (Present),Bacteria (Not Present),Bacteria (Present),Hypertension (No),Hypertension (Yes),Diabetes Mellitus (No),Diabetes Mellitus (Yes),Coronary Artery Disease (No),Coronary Artery Disease (Yes),Appetite (Good),Appetite (Poor),Pedal Edema (No),Pedal Edema (Yes),Anemia (No),Anemia (Yes),Target,Target (Not CKD)
0,0.796206,0.723662,1.137875,3.951101,1.324211,-0.934242,1.472513,-2.795758,-2.712933,0.506560,-1.790290,2,0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,0,1,0,1,0,1,1,0
1,0.921652,2.765149,3.506466,-0.236745,1.324211,0.456741,-1.090042,-0.360450,-0.528732,0.770681,-0.679647,2,2,0,1,0,1,1,0,0,1,0,1,1,0,0,1,1,0,1,0,1,0,1,0
2,1.548881,0.723662,2.908784,3.728738,1.974189,-0.412624,1.472513,-1.578104,-1.484320,-0.373843,-0.679647,2,0,1,0,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,1,0,1,0
3,-1.712709,1.744405,-0.323125,-0.051442,0.024253,-2.672971,-1.090042,-2.389874,-2.849446,1.915205,-1.790290,4,0,0,1,1,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,0,1,1,0
4,0.984375,2.765149,0.916512,0.467406,3.924126,-0.064878,0.191235,-2.795758,-2.439908,-0.241782,-1.790290,4,2,1,0,1,0,1,0,0,1,0,1,0,1,1,0,1,0,0,1,1,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
129,-1.775432,-0.297082,0.031057,0.096800,0.024253,-0.934242,-1.090042,0.045435,0.017319,-1.122185,-0.679647,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
130,-1.649986,-1.317826,-0.544488,-0.866775,0.024253,-0.412624,0.191235,-0.360450,-0.255706,-0.065702,1.541638,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
131,-0.270083,-1.317826,-0.588761,0.170921,-0.625726,-0.412624,0.191235,0.451320,0.836394,-0.153742,1.541638,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1
132,0.608037,0.723662,-0.478079,0.319164,0.024253,-0.064878,-1.090042,0.045435,0.836394,-0.593943,1.541638,0,0,0,1,0,1,1,0,1,0,1,0,1,0,1,0,1,0,1,0,1,0,0,1


## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [55]:
# Converted SQL to Pandas code
result = final_data.groupby('Target').size().reset_index(name='count')

# Display the result
print(result)

   Target  count
0       0    111
1       1     23


The code above counts the amount of patients that have ckd and don't have ckd and places the values in a table.